<a href="https://colab.research.google.com/github/ShakawatBhai/Phitron_All_ML_DL_Projects/blob/main/DL_Assignment_03_Question_solved.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# DL Assignment 03

**Name:** Mohammed Shakawat Hossen

**Course Email:** shakawathosen156@gmail.com

Dataset : https://www.kaggle.com/datasets/uciml/breast-cancer-wisconsin-data


## End of Assignment

Before submitting:
- Run all cells from top to bottom.  
- Check that all answer sections are filled.  
- Instruction video অনুযায়ী আমাদের দেয়া Colab ফাইলটি থেকে প্রথম একটি Save copy in drive করে নিবা। এরপর Google colab এর মধ্যে কোডগুলো করবে এবং সেই ফাইলটি ‘Anyone with the link’ & ‘View’ Access দিয়ে ফাইলটির Shareble Link টি সাবমিট করবে।

# General Instruction

You must choose your own dataset.

The dataset must:

Be a supervised learning dataset (Regression or Binary Classification)

Contain at least 300 samples

Have at least 2 input features

Be in CSV format

You are NOT allowed to use Dataset or DataLoader.

You must implement everything manually.

# Question 01: [ Marks 05 ]

## Dataset Preparation

## Using your chosen dataset:

Load the dataset.

Perform necessary preprocessing:

Handle missing values (if any)

Encode categorical variables (if necessary)

Feature scaling (if needed)

Separate features (X) and target (y).

Convert them into NumPy arrays.

Convert them into PyTorch tensors.

Split into training and testing sets.

Clearly explain each preprocessing decision.

# **Write** Answer 01:


In [ ]:
import numpy as np
import pandas as pd

df = pd.read_csv("data.csv")
df.head()

,id,diagnosis,radius_mean,texture_mean,perimeter_mean,area_mean,smoothness_mean,compactness_mean,concavity_mean,concave points_mean,...,texture_worst,perimeter_worst,area_worst,smoothness_worst,compactness_worst,concavity_worst,concave points_worst,symmetry_worst,fractal_dimension_worst,Unnamed: 32
0,842302,M,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.3001,0.14710,...,17.33,184.60,2019.0,0.1622,0.6656,0.7119,0.2654,0.4601,0.11890,NaN
1,842517,M,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.0869,0.07017,...,23.41,158.80,1956.0,0.1238,0.1866,0.2416,0.1860,0.2750,0.08902,NaN
2,84300903,M,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.1974,0.12790,...,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758,NaN
3,84348301,M,11.42,20.38,77.58,386.1,0.14250,0.28390,0.2414,0.10520,...,26.50,98.87,567.7,0.2098,0.8663,0.6869,0.2575,0.6638,0.17300,NaN
4,84358402,M,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.1980,0.10430,...,16.67,152.20,1575.0,0.1374,0.2050,0.4000,0.1625,0.2364,0.07678,NaN


In [ ]:
df = df.drop(columns=["id", "Unnamed: 32"])
print(df.isnull().sum())

diagnosis                  0
radius_mean                0
texture_mean               0
perimeter_mean             0
area_mean                  0
smoothness_mean            0
compactness_mean           0
concavity_mean             0
concave points_mean        0
symmetry_mean              0
fractal_dimension_mean     0
radius_se                  0
texture_se                 0
perimeter_se               0
area_se                    0
smoothness_se              0
compactness_se             0
concavity_se               0
concave points_se          0
symmetry_se                0
fractal_dimension_se       0
radius_worst               0
texture_worst              0
perimeter_worst            0
area_worst                 0
smoothness_worst           0
compactness_worst          0
concavity_worst            0
concave points_worst       0
symmetry_worst             0
fractal_dimension_worst    0
dtype: int64


In [ ]:
df["diagnosis"] = df["diagnosis"].map({"M": 0, "B": 1})
X = df.drop(columns=["diagnosis"])
y = df["diagnosis"]

X = X.values
y = y.values.reshape(-1, 1)

X = (X-X.mean(axis=0))/X.std(axis=0)

split = int(0.8*len(X))

X_train = X[:split]
y_train = y[:split]

X_test = X[split:]
y_test = y[split:]



In [ ]:
import torch

X_train = torch.tensor(X_train,dtype=torch.float32)
y_train = torch.tensor(y_train,dtype=torch.float32)

X_test = torch.tensor(X_test,dtype=torch.float32)
y_test = torch.tensor(y_test,dtype=torch.float32)


Removed columns (id, Unnamed: 32) → not useful for prediction

Checked missing values → none found, so no action needed

Encoded target (M, B) → (0, 1) → models need numeric data

Separated features (X) and target (y) → required for supervised learning

Converted to NumPy arrays → efficient computation

Applied feature scaling → ensures all features contribute equally

Split data (80% train, 20% test) → evaluate model on unseen data

Converted to PyTorch tensors → required for model training

# Question 02: [ Marks 20 ]

## Design a neural network using nn.Module.

### The model must contain:

Input layer

At least one hidden layer

Output layer

Suitable activation function



## Justify:

Number of hidden neurons

Choice of activation function

Print  the total number of trainable parameters.


## Write Answer 02:


In [ ]:
import torch.nn as nn

class BreastCancerNN(nn.Module):
    def __init__(self, input_size):
        super(BreastCancerNN, self).__init__()

        self.fc1 = nn.Linear(input_size, 16)
        self.relu = nn.ReLU()
        self.fc2 = nn.Linear(16, 1)
        self.sigmoid = nn.Sigmoid()

    def forward(self, x):
        x = self.fc1(x)
        x = self.relu(x)
        x = self.fc2(x)
        x = self.sigmoid(x)
        return x


In [ ]:
model = BreastCancerNN(input_size=X_train.shape[1])
print(model)


BreastCancerNN(
  (fc1): Linear(in_features=30, out_features=16, bias=True)
  (relu): ReLU()
  (fc2): Linear(in_features=16, out_features=1, bias=True)
  (sigmoid): Sigmoid()
)


In [ ]:
total_params = sum(p.numel() for p in model.parameters() if p.requires_grad)
print("Total Trainable Parameters:", total_params)


Total Trainable Parameters: 513


1. Number of Hidden Neurons (16)
Dataset is moderate (30 features, ~569 samples)

16 neurons:
Enough to learn patterns

Avoids overfitting
Rule of thumb: between input size and output size

2. Activation Function

ReLU (Hidden Layer)
Fast and simple
Avoids vanishing gradient problem
Works well for deep learning

Sigmoid (Output Layer)
Used for binary classification
Outputs probability (0–1)

# Question 03: [ Marks 10 ]

Choose an appropriate loss function.

Choose an optimizer.

<br>

Justify your choices based on:

Regression vs Classification

Nature of the dataset

## Write Answer 03:

In [ ]:
criterion = nn.BCELoss()


The problem is Binary Classification (Malignant vs Benign)

Output is a probability (0–1) using Sigmoid

BCELoss measures error between predicted probability and true label

In [ ]:
optimizer = torch.optim.Adam(model.parameters(), lr=0.001)


Efficient and fast convergence

Automatically adjusts learning rate

Works well for most neural network problems

Since the dataset is a binary classification problem, BCELoss is appropriate because it works with probability outputs.
The dataset is medium-sized with multiple features, so Adam optimizer is chosen for stable and faster training compared to basic gradient descent.

# Question 04: [ Marks 15 ]

## Implement a full training loop:

Forward pass

Loss computation

Backward pass

Parameter update

Gradient reset

### Requirements:

Train for at least 100 epochs.

Print loss every 10 epochs.

Store training loss history(You can pick your own Data Structure).

Explain clearly what happens in each step of the pipeline.

## Write Answer 04:

In [ ]:
num_epochs = 100
loss_history = []

for epoch in range(num_epochs):

    outputs = model(X_train)

    loss = criterion(outputs, y_train)

    loss.backward()

    optimizer.step()

    optimizer.zero_grad()

    loss_history.append(loss.item())

    if (epoch + 1) % 10 == 0:
        print(f"Epoch [{epoch+1}/{num_epochs}], Loss: {loss.item():.4f}")


Epoch [10/100], Loss: 0.6370
Epoch [20/100], Loss: 0.5672
Epoch [30/100], Loss: 0.5052
Epoch [40/100], Loss: 0.4471
Epoch [50/100], Loss: 0.3931
Epoch [60/100], Loss: 0.3443
Epoch [70/100], Loss: 0.3018
Epoch [80/100], Loss: 0.2657
Epoch [90/100], Loss: 0.2355
Epoch [100/100], Loss: 0.2104


Forward pass: Model takes input and produces predictions

Loss computation: Calculates error between prediction and actual value

Backward pass: Computes gradients (how to fix the error)

Parameter update: Optimizer updates weights to reduce loss

Gradient reset: Clears old gradients to avoid accumulation

Loss history: Stores loss to track learning progress

# Question 05: [ Marks 10 ]

## Evaluate the model on test data.

## For regression:

Report MSE and MAE


## For classification:

Report Accuracy

Compare training vs testing performance.

State whether the model is underfitting or overfitting.

## Write Answer 05:

In [ ]:
with torch.no_grad():
    y_pred = model(X_test)

    y_pred_label = (y_pred>0.5).float()

    accuracy = (y_pred_label == y_test).sum().item() / len(y_test)

print("Test Accuracy:", accuracy)


Test Accuracy: 0.9298245614035088


In [ ]:
with torch.no_grad():
    train_pred = model(X_train)
    train_pred_label = (train_pred > 0.5).float()

    train_accuracy = (train_pred_label == y_train).sum().item() / len(y_train)

print("Train Accuracy:", train_accuracy)


Train Accuracy: 0.9560439560439561


The model achieved 95.60% training accuracy and 92.98% testing accuracy. The small difference between them indicates slight overfitting, but overall the model generalizes well and performs effectively on unseen data.


# Question 06: [ Marks 20 ]

## Modify at least ONE of the following:

Learning rate

Number of hidden neurons

Number of epochs

### Train again and compare:

Convergence speed

Final performance

Explain how the change affected the model.

## Write Answer 06:

In [ ]:
#modified(Hidden neurons: 32)

class ModifiedNN(nn.Module):
    def __init__(self, input_size):
        super(ModifiedNN, self).__init__()

        self.fc1 = nn.Linear(input_size, 32)  # changed from 16 → 32
        self.relu = nn.ReLU()
        self.fc2 = nn.Linear(32, 1)
        self.sigmoid = nn.Sigmoid()

    def forward(self, x):
        x = self.fc1(x)
        x = self.relu(x)
        x = self.fc2(x)
        x = self.sigmoid(x)
        return x


In [ ]:
model = ModifiedNN(input_size=X_train.shape[1])

criterion = nn.BCELoss()
optimizer = torch.optim.Adam(model.parameters(), lr=0.001)


In [ ]:
num_epochs = 100
loss_history_new = []

for epoch in range(num_epochs):

    outputs = model(X_train)
    loss = criterion(outputs, y_train)

    loss.backward()
    optimizer.step()
    optimizer.zero_grad()

    loss_history_new.append(loss.item())

    if (epoch + 1) % 10 == 0:
        print(f"Epoch [{epoch+1}/100], Loss: {loss.item():.4f}")


Epoch [10/100], Loss: 0.6365
Epoch [20/100], Loss: 0.5426
Epoch [30/100], Loss: 0.4612
Epoch [40/100], Loss: 0.3889
Epoch [50/100], Loss: 0.3275
Epoch [60/100], Loss: 0.2775
Epoch [70/100], Loss: 0.2379
Epoch [80/100], Loss: 0.2067
Epoch [90/100], Loss: 0.1822
Epoch [100/100], Loss: 0.1626


In [ ]:
with torch.no_grad():
    y_pred = model(X_test)
    y_pred_label = (y_pred > 0.5).float()

    test_acc_new = (y_pred_label == y_test).sum().item() / len(y_test)

print("New Test Accuracy:", test_acc_new)


New Test Accuracy: 0.9649122807017544


Increasing the number of hidden neurons from 16 to 32 improved the model’s capacity to learn complex patterns, resulting in better feature representation and higher test accuracy. This indicates that the previous model was slightly underfitting, and the increased neurons helped achieve better generalization without causing overfitting.

# Question 07: [ Marks 20 ]


# Training Analysis

Answer the following:

Why must gradients be reset every epoch?

What happens if learning rate is too high?

What happens if learning rate is too small?

Why do we define layers inside the constructor (__init__) and not inside forward()?


## Write Answer 07:

Why must gradients be reset every epoch?

Gradients accumulate. Resetting them prevents previous epoch's gradients from influencing current updates, ensuring correct learning.

What happens if the learning rate is too high?

Too high a learning rate causes the model to overshoot the minimum, bounce erratically, or diverge, leading to unstable training and increased loss.

What happens if the learning rate is too small?

Too small a learning rate makes training very slow, potentially trapping the model in local minima or plateaus, and delaying convergence significantly.

Why do we define layers inside the constructor (__init __) and not inside forward()?

Layers are model components with learnable parameters, defined once in __init __ for initialization. forward() defines the data flow through these already-initialized layers; defining layers there would re-create them repeatedly, preventing learning and wasting memory.